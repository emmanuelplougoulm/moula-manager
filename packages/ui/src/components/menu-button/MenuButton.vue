<template>
  <GroupButtonContainer :type="type" :disabled="disabled" @click="click">
    <template #left-icon>
      <SvgIcon :value="leftIcon" class="left-icon" />
    </template>
    <template #text>
      <span class="title text-default-bold" :class="textClass">
        {{ title }}
      </span>
      <span v-if="subtitle" class="subtitle text-small-regular" :class="textClass">
        {{ subtitle }}
      </span>
    </template>
    <template #right-icon>
      <SvgIcon :value="rightIcon" class="right-icon" />
    </template>
  </GroupButtonContainer>
</template>

<script setup lang="ts">
import GroupButtonContainer from '../group-button-container/GroupButtonContainer.vue';
import { iconValidator, typeValidator } from './validator';
import SvgIcon from '../icons/SvgIcon.vue';
import { computed } from 'vue';

const props = defineProps({
  type: {
    type: String,
    required: true,
    validator: typeValidator
  },
  leftIcon: {
    type: String,
    required: true,
    validator: iconValidator
  },
  rightIcon: {
    type: String,
    required: true,
    validator: iconValidator
  },
  title: {
    type: String,
    required: true
  },
  subtitle: {
    type: String,
    default: undefined
  },
  disabled: {
    type: Boolean,
    default: false
  }
});

const textClass = computed(() => ({
  ['text--highlighted']: props.type === 'highlighted' || props.type === 'completed'
}));

const emit = defineEmits(['click']);
const click = () => {
  if (!props.disabled) {
    emit('click');
  }
};
</script>

<style scoped>
.title {
  color: var(--neutral-20);
}

.subtitle {
  color: var(--neutral-30);
}

.text--highlighted {
  color: var(--primary-text);
}

.left-icon {
  height: 2.5rem;
  width: 2.5rem;
}

.right-icon {
  width: 1.125rem;
  height: 1.125rem;
}
</style>
