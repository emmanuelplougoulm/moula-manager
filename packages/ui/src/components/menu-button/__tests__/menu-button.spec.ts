import Icons from '../../icons/index';
import { mount } from '@vue/test-utils';
import MenuButton from '../MenuButton.vue';

const button = "[data-testid='button']";

describe('MenuButton tests', () => {
  it('should display the component with the default type', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'ArrowForwardIOS',
        type: 'default'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with the ghost type', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'ArrowForwardIOS',
        type: 'ghost'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with the highlighted type', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'ArrowForwardIOS',
        type: 'highlighted'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with the completed type', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'MoreVert',
        type: 'completed'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with the disabled prop set to true', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'MoreVert',
        type: 'completed',
        disabled: true
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component without the subtitle', () => {
    const wrapper = mount(MenuButton, {
      global: {
        plugins: [Icons]
      },
      props: {
        title: 'Title',
        leftIcon: 'Plus',
        rightIcon: 'ArrowForwardIOS',
        type: 'default'
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit a click event when the button is clicked', async () => {
    const wrapper = mount(MenuButton, {
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'MoreVert',
        type: 'completed'
      }
    });
    await wrapper.find(button).trigger('click');
    expect(wrapper.emitted()).toHaveProperty('click');
  });

  it('should not emit a click event when the button is clicked if the disabled prop is set to true', async () => {
    const wrapper = mount(MenuButton, {
      props: {
        title: 'Title',
        subtitle: 'Subtitle',
        leftIcon: 'Plus',
        rightIcon: 'MoreVert',
        type: 'completed',
        disabled: true
      }
    });
    await wrapper.find(button).trigger('click');
    expect(wrapper.emitted()).not.toHaveProperty('click');
  });
});
