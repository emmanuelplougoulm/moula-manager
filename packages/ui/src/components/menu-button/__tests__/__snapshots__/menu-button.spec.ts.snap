// Vitest Snapshot v1

exports[`MenuButton tests > should display the component with the completed type 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button button--completed\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold text--highlighted\\">Title</span><span data-v-44ea4910=\\"\\" class=\\"subtitle text-small-regular text--highlighted\\">Subtitle</span></div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"right-icon--completed\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"MoreVert\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480.281 915.218q-22.542 0-38.52-15.968-15.979-15.968-15.979-38.391 0-21.893 15.927-38.267 15.927-16.374 38.293-16.374 22.824 0 38.52 16.445 15.696 16.445 15.696 38.522 0 22.077-15.698 38.055-15.697 15.978-38.239 15.978Zm0-285q-22.542 0-38.52-15.927-15.979-15.927-15.979-38.293 0-22.824 15.927-38.52 15.927-15.696 38.293-15.696 22.824 0 38.52 15.698 15.696 15.697 15.696 38.239 0 22.542-15.698 38.52-15.697 15.979-38.239 15.979Zm0-284.436q-22.542 0-38.52-16.14-15.979-16.14-15.979-38.804 0-22.665 15.927-38.643 15.927-15.979 38.293-15.979 22.824 0 38.52 16.12 15.696 16.12 15.696 38.785 0 22.665-15.698 38.663-15.697 15.998-38.239 15.998Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;

exports[`MenuButton tests > should display the component with the default type 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold\\">Title</span><span data-v-44ea4910=\\"\\" class=\\"subtitle text-small-regular\\">Subtitle</span></div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"ArrowForwardIOS\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;

exports[`MenuButton tests > should display the component with the disabled prop set to true 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button button--completed\\" disabled=\\"\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold text--highlighted\\">Title</span><span data-v-44ea4910=\\"\\" class=\\"subtitle text-small-regular text--highlighted\\">Subtitle</span></div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"right-icon--completed\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"MoreVert\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M480.281 915.218q-22.542 0-38.52-15.968-15.979-15.968-15.979-38.391 0-21.893 15.927-38.267 15.927-16.374 38.293-16.374 22.824 0 38.52 16.445 15.696 16.445 15.696 38.522 0 22.077-15.698 38.055-15.697 15.978-38.239 15.978Zm0-285q-22.542 0-38.52-15.927-15.979-15.927-15.979-38.293 0-22.824 15.927-38.52 15.927-15.696 38.293-15.696 22.824 0 38.52 15.698 15.696 15.697 15.696 38.239 0 22.542-15.698 38.52-15.697 15.979-38.239 15.979Zm0-284.436q-22.542 0-38.52-16.14-15.979-16.14-15.979-38.804 0-22.665 15.927-38.643 15.927-15.979 38.293-15.979 22.824 0 38.52 16.12 15.696 16.12 15.696 38.785 0 22.665-15.698 38.663-15.697 15.998-38.239 15.998Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;

exports[`MenuButton tests > should display the component with the ghost type 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button button--ghost\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold\\">Title</span><span data-v-44ea4910=\\"\\" class=\\"subtitle text-small-regular\\">Subtitle</span></div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"ArrowForwardIOS\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;

exports[`MenuButton tests > should display the component with the highlighted type 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button button--highlighted\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold text--highlighted\\">Title</span><span data-v-44ea4910=\\"\\" class=\\"subtitle text-small-regular text--highlighted\\">Subtitle</span></div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"ArrowForwardIOS\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;

exports[`MenuButton tests > should display the component without the subtitle 1`] = `
"<div data-v-d42ab394=\\"\\" data-v-44ea4910=\\"\\" class=\\"container\\"><button data-v-d42ab394=\\"\\" type=\\"button\\" class=\\"button\\" data-testid=\\"button\\">
    <div data-v-d42ab394=\\"\\" class=\\"icon-text-container\\">
      <div data-v-d42ab394=\\"\\" class=\\"icon-container\\"><svg data-v-44ea4910=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"40\\" height=\\"40\\" fill=\\"none\\" alt=\\"Plus\\" class=\\"left-icon\\">
          <path fill=\\"currentColor\\" d=\\"M32.444 18.75H21.25V7.557a1.25 1.25 0 1 0-2.5 0V18.75H7.557a1.25 1.25 0 1 0 0 2.5H18.75v11.194a1.25 1.25 0 1 0 2.5 0V21.25h11.194a1.25 1.25 0 1 0 0-2.5Z\\"></path>
        </svg></div>
      <div data-v-d42ab394=\\"\\" class=\\"text-container\\"><span data-v-44ea4910=\\"\\" class=\\"title text-default-bold\\">Title</span>
        <!--v-if-->
      </div>
    </div>
    <div data-v-d42ab394=\\"\\" class=\\"\\">
      <div data-v-b9dd0015=\\"\\" data-v-44ea4910=\\"\\" class=\\"container right-icon\\" alt=\\"ArrowForwardIOS\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
          <path d=\\"M304 987.566 234.434 917l343-343-343-343L304 160.434 717.566 574 304 987.566Z\\"></path>
        </svg></div>
    </div>
  </button></div>"
`;
